src/rules/infinite-reactive-loop.ts

```diff
@@ -0,0 +1,439 @@
+
+/** Let's lint! */
+function doLint(
+  context: RuleContext,
+  ast: TSESTree.Node,
+  callFuncIdentifiers: TSESTree.Identifier[],
+  tickCallExpressions: { node: TSESTree.CallExpression; name: string }[],
+  taskReferences: {
+    node: TSESTree.CallExpression
+    name: string
+  }[],
+  reactiveVariableNames: string[],
+  pIsSameTask: boolean,
+) {
+  let isSameMicroTask = pIsSameTask
+
+  traverseNodes(ast, {
+    enterNode(node) {
+      // Promise.then() or Promise.catch() is called.
+      if (isPromiseThenOrCatchBody(node)) {
+        isSameMicroTask = false
+      }
+
+      // `tick`, `setTimeout`, `setInterval` , `queueMicrotask` is called
+      for (const { node: callExpression } of [
+        ...tickCallExpressions,
+        ...taskReferences,
+      ]) {
+        if (isChildNode(callExpression, node)) {
+          isSameMicroTask = false
+        }
+      }
+
+      // left side of await block
+      if (
+        node.parent?.type === "AssignmentExpression" &&
+        node.parent?.right.type === "AwaitExpression" &&
+        node.parent?.left === node
+      ) {
+        isSameMicroTask = false
+      }
+
+      if (node.type === "Identifier" && isFunctionCall(node)) {
+        // traverse used functions body
+        const functionDeclarationNode = getFunctionDeclarationNode(node)
+        if (functionDeclarationNode) {
+          doLint(
+            context,
+            functionDeclarationNode,
+            [...callFuncIdentifiers, node],
+            tickCallExpressions,
+            taskReferences,
+            reactiveVariableNames,
+            isSameMicroTask,
+          )
+        }
+      }
+
+      if (!isSameMicroTask) {
+        if (
+          isReactiveVariableNode(context, node) &&
+          reactiveVariableNames.includes(node.name) &&
+          isNodeForAssign(node)
+        ) {
+          context.report({
+            node,
+            loc: node.loc,
+            messageId: "unexpected",
+          })
+          callFuncIdentifiers.forEach((callFuncIdentifier) => {
+            context.report({
+              node: callFuncIdentifier,
+              loc: callFuncIdentifier.loc,
+              messageId: "unexpectedCall",
+              data: {
+                variableName: node.name,
+              },
+            })
+          })
+        }
+      }
+    },
+    leaveNode(node) {
+      if (node.type === "AwaitExpression") {
+        if ((ast.parent?.type as string) === "SvelteReactiveStatement") {
+          // MEMO: It checks that `await` is used in reactive statement directly or not.
+          // If `await` is used in inner function of a reactive statement, result of `isInsideOfFunction` will be `true`.
+          if (!isInsideOfFunction(node)) {
+            isSameMicroTask = false
+          }
+        } else {
+          isSameMicroTask = false
+        }
+      }
+
+      // Promise.then() or Promise.catch() is called.
+      if (isPromiseThenOrCatchBody(node)) {
+        isSameMicroTask = true
```

Comment:
I think we can remove the duplicate code by keeping the `node` when it changes to `isSameMicroTask = false` in `enterNode()` so that when the same `node` comes, `isSameMicroTask = true`. What do you think?

Reply:
Good catch. I removed the duplicate code.

https://github.com/ota-meshi/eslint-plugin-svelte/pull/332/commits/336dfa1949b39f6729b7ba3b2aaf63c0ce66b62e